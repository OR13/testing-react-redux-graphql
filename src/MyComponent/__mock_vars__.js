export const getPokemon = {
  name: "Charmander"
};
