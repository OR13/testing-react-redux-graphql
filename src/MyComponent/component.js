import React from "react";

export default class MyComponent extends React.Component {
  render() {
    return (
      <div style={{ padding: "20px" }}>
        <h4> Pokemon from GraphQL</h4>

        <h5>pokemon from props</h5>
        <pre>
          {this.props.data &&
            this.props.data.pokemon &&
            this.props.data.pokemon.name}
        </pre>

        <h5>pokemon from redux</h5>
        <pre>
          {this.props.myPokedex &&
            this.props.myPokedex.selectedPokemon &&
            this.props.myPokedex.selectedPokemon.name}
        </pre>

        <button
          onClick={() => {
            this.props.loadPokemonWithApollo({ name: "Charmander" });
          }}
        >
          load pokemon
        </button>
      </div>
    );
  }
}
