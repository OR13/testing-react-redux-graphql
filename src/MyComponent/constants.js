export const POKEMON_LOAD_STARTED = "POKEMON_LOAD_STARTED";

export const POKEMON_LOADED = "POKEMON_LOADED";

export const POKEMON_LOAD_FAILED = "POKEMON_LOAD_FAILED";
