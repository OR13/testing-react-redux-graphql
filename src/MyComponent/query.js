import gql from "graphql-tag";

export const getPokemon = gql`
    query getPokemonByName($name: String!){
      pokemon(name: $name) {
      id
      number
      name
      attacks {
        special {
          name
          type
          damage
        }
      }
      evolutions {
        id
        number
        name
        weight {
          minimum
          maximum
        }
        attacks {
          fast {
            name
            type
            damage
          }
        }
      }
    }
  }
`;
