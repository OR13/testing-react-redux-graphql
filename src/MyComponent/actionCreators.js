import * as Constants from "./constants";

export const pokemonLoadStarted = () => {
  return {
    type: Constants.POKEMON_LOAD_STARTED
  };
};

export const pokemonLoaded = pokemon => {
  return {
    type: Constants.POKEMON_LOADED,
    payload: pokemon
  };
};

export const pokemonLoadFailed = error => {
  return {
    type: Constants.POKEMON_LOAD_FAILED,
    payload: {
      error
    }
  };
};
