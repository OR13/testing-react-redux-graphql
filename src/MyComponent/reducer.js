import * as Constants from "./constants";


const handlers = {
  [Constants.POKEMON_LOAD_STARTED]: (state, action) => {
    return {
      ...state,
      status: 'PENDING'
    };
  },
  [Constants.POKEMON_LOADED]: (state, action) => {
    return {
      ...state,
      status: 'SUCCESS',
      selectedPokemon: action.payload
    };
  },
  [Constants.POKEMON_LOAD_FAILED]: (state, action) => {
    return {
      ...state,
      status: 'FAILED'
    };
  },
};

export default (state, action) => {
  if (handlers[action.type]) {
    return handlers[action.type](state, action);
  }
  return {
    ...state
  };
};
