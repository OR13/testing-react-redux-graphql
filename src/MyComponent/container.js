import { graphql, compose } from "react-apollo";

import { getPokemon } from "./query";

import { connect } from "react-redux";
import client from "../Apollo";

import {
  pokemonLoadStarted,
  pokemonLoaded,
  pokemonLoadFailed
} from "./actionCreators";

const mapStateToProps = state => {
  return {
    myPokedex: state.pokedex
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onPokemonLoadStarted: () => {
      dispatch(pokemonLoadStarted());
    },
    onPokemonLoaded: pokemon => {
      dispatch(pokemonLoaded(pokemon));
    },
    onPokemonLoadFailed: () => {
      dispatch(pokemonLoadFailed());
    }
  };
};

export const gqlContainer = component => {
  return compose(
    graphql(getPokemon, {
      options: ({ name }) => ({
        variables: {
          name
        }
      })
    }),
    graphql(getPokemon, {
      props: ({ ownProps }) => ({
        loadPokemonWithApollo: async model => {
          ownProps.onPokemonLoadStarted();
          try {
            const result = await client.query({
              query: getPokemon,
              variables: model
            });
            ownProps.onPokemonLoaded(result.data.pokemon);
            return result.data.pokemon;
          } catch (e) {
            console.log(e);
            ownProps.onPokemonLoadFailed(e);
          }
        }
      })
    })
  )(component);
};

export const reduxContainer = component => {
  const hoc = gqlContainer(component);
  return connect(mapStateToProps, mapDispatchToProps)(hoc);
};
