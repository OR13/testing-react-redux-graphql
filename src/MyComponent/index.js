import { reduxContainer } from "./container";

import MyComponent from "./component";

export default reduxContainer(MyComponent);
