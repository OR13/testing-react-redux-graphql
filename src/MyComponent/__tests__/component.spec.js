import React from "react";
import ReactDOM from "react-dom";
import MyComponent from "../component";

import { shallow } from "enzyme";
import toJson from 'enzyme-to-json'

it("component does not crash", () => {
  const component = shallow(<MyComponent />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
