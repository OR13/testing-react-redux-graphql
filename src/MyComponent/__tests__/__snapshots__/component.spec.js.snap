// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component does not crash 1`] = `
<div
  style={
    Object {
      "padding": "20px",
    }
  }
>
  <h4>
     Pokemon from GraphQL
  </h4>
  <h5>
    pokemon from props
  </h5>
  <pre />
  <h5>
    pokemon from redux
  </h5>
  <pre />
  <button
    onClick={[Function]}
  >
    load pokemon
  </button>
</div>
`;
